File: src/main/java/com/dude/dms/ui/components/RegexField.java
package com.dude.dms.ui.components;

import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.textfield.TextField;

public class RegexField extends HorizontalLayout {

    private final TextField regex;

    private ValueProvider<String> reference;

    public RegexField() {
        this("");
    }

    public RegexField(String label) {
        this(label, "");
    }

    public RegexField(String label, String initialValue) {
        regex = new TextField(label, initialValue, "");
        regex.setSizeFull();
        Icon icon = VaadinIcon.EXPAND.create();
        icon.getElement().getStyle().set("paddingBottom", "8px");
        icon.addClickListener(event -> {
            if (reference != null) {
                new RegexDialog(regex, reference).open();
            } else {
                new RegexDialog(regex).open();
            }
        });

        setPadding(false);
        setMargin(false);
        setAlignItems(Alignment.END);
        add(icon, regex);
    }

    public void setValue(String value) {
        regex.setValue(value);
    }

    public String getValue() {
        return regex.getValue();
    }

    public void setReference(ValueProvider<String> reference) {
        this.reference = reference;
    }

    public boolean isEmpty() {
        return regex.isEmpty();
    }
}

File: src/main/java/com/dude/dms/ui/components/standard/RegexField.kt
package com.dude.dms.ui.components.standard

import com.dude.dms.extensions.regexDialog
import com.github.mvysny.karibudsl.v10.icon
import com.github.mvysny.karibudsl.v10.onLeftClick
import com.github.mvysny.karibudsl.v10.textField
import com.vaadin.flow.component.icon.VaadinIcon
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.textfield.TextField

class RegexField(label: String = "") : HorizontalLayout() {

    private lateinit var regex: TextField

    var value: String?
        get() = regex.value
        set(value) {
            regex.value = value
        }

    val isEmpty: Boolean
        get() = regex.isEmpty

    init {
        isPadding = false
        isMargin = false
        alignItems = FlexComponent.Alignment.END

        icon(VaadinIcon.EXPAND) {
            style["paddingBottom"] = "8px"
            onLeftClick { regexDialog(regex).open() }
        }
        regex = textField(label) { setSizeFull() }
    }
}